cmd/rsample/main.go

```diff
@@ -0,0 +1,209 @@
+	"fmt"
+	"log"
+	"math"
+	"os"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/go-redis/redis/v8"
+)
+
+type keyspace struct {
+	shard   string
+	keys    int
+	expires int
+	avgTtl  int
+}
+
+type (
+	histogram struct {
+		name string
+		v    []int64 // sorted
+	}
+	histogramSlice []*histogram
+)
+
+func (h *histogram) Count() int                 { return len(h.v) }
+func (h *histogram) Min() int64                 { return h.v[0] }
+func (h *histogram) Max() int64                 { return h.v[len(h.v)-1] }
+func (h *histogram) Percentile(p float64) int64 { return h.v[int(math.Floor(p*float64(len(h.v))))] }
+
+func (hs histogramSlice) Len() int           { return len(hs) }
+func (hs histogramSlice) Swap(i, j int)      { hs[i], hs[j] = hs[j], hs[i] }
+func (hs histogramSlice) Less(i, j int) bool { return hs[i].Count() > hs[j].Count() }
+
+func main() {
+	urlList := flag.String("swarm-redis-urls", "", "Redis URLs as comma separated list")
+	samples := flag.Int("samples", 0, "Number of random keys to check")
+	threads := flag.Int("threads", 1, "Number of threads to use for sampling")
+	flag.Parse()
+
+	urls := strings.Split(*urlList, ",")
+	if urls[0] == "" || *samples < 0 || *threads < 1 {
+		usage()
+	}
+
+	ctx := context.Background()
+	ring := newRing(urls)
+	defer ring.Close()
+
+	keyspaces, err := keyspaces(ctx, ring)
+	if err != nil {
+		log.Fatalln(err)
+	}
+	histograms := measure(ctx, ring, *samples, *threads)
+
+	report(keyspaces, histograms)
+}
+
+func usage() {
+	fmt.Fprintf(flag.CommandLine.Output(), "Usage of %s:\n", os.Args[0])
+	flag.PrintDefaults()
+	os.Exit(1)
+}
+
+func newRing(urls []string) *redis.Ring {
+	opts := &redis.RingOptions{
+		Addrs: map[string]string{},
+	}
+	for idx, addr := range urls {
+		opts.Addrs[fmt.Sprintf("redis%d", idx)] = addr
+	}
+	ring := redis.NewRing(opts)
+	return ring
+}
+
+func keyspaces(ctx context.Context, ring *redis.Ring) (result []keyspace, err error) {
+	err = ring.ForEachShard(ctx, func(ctx context.Context, client *redis.Client) error {
+		info, err := client.Info(ctx, "keyspace").Result()
+		if err == nil {
+			// # Keyspace
+			// db0:keys=1,expires=1,avg_ttl=999336
+			var ks keyspace
+			_, err := fmt.Sscanf(info, "# Keyspace\ndb0:keys=%d,expires=%d,avg_ttl=%d", &ks.keys, &ks.expires, &ks.avgTtl)
+			if err != nil {
+				return err
+			}
+			ks.shard = client.Options().Addr
+			result = append(result, ks)
+		}
+		return err
+	})
+	return
+}
+
+func measure(ctx context.Context, ring *redis.Ring, samples int, threads int) (hs histogramSlice) {
+	var maps []map[string][]int64
+	var wg sync.WaitGroup
+
+	for i := 0; i < threads; i++ {
+		m := make(map[string][]int64)
+		maps = append(maps, m)
+
+		wg.Add(1)
+		go func() {
+			for j := 0; j < samples/threads; j++ {
+				key := ring.RandomKey(ctx).Val()
+
+				group, ok := group(key)
+				if !ok {
+					// ignore unrecognized keys
+					continue
+				}
+
+				count, err := ring.ZCard(ctx, key).Result()
+				if err != nil {
+					continue
+				}
```

Comment:
Use `func getSample() (name string, count int64, err error)` parameter. Also makes this function testable